### PR TITLE
feat: Add comprehensive image hashing functions (perceptual, difference, wavelet, crop-resistant)

### PR DESCRIPTION
**Pull request by codekshitij**
# Changes Made

This PR adds 5 comprehensive image hashing functions to Daft's image processing capabilities, enabling robust image similarity detection and deduplication.

## New Functions Added:
Average Hash (aHash) - df["images"].image.average_hash()
Perceptual Hash (pHash) - df["images"].image.perceptual_hash()
Difference Hash (dHash) - df["images"].image.difference_hash()
Wavelet Hash (wHash) - df["images"].image.wavelet_hash()
Crop-Resistant Hash (cHash) - df["images"].image.crop_resistant_hash()

## Implementation Details:

Rust Backend (src/daft-image/):
Added 5 new ScalarUDF structs in functions/hash.rs

Implemented core hash computation algorithms in ops.rs:
Average Hash: Compares pixel averages to global average
Perceptual Hash: Uses DCT transform for robust similarity detection
Difference Hash: Compares adjacent pixels horizontally (9x8 → 8x8)
Wavelet Hash: Uses Haar wavelet transform (64x64 → 8x8 coefficients)
Crop-Resistant Hash: Analyzes 16 overlapping 64x64 regions on 256x256 image
Added series-level functions in series.rs with proper type dispatch
Registered all functions in the global function registry

Python API (daft/expressions/expressions.py):
Added methods to ImageExpression class with comprehensive docstrings
All functions return 64-character binary strings for consistency
Support for both Image and FixedShapeImage data types
Proper error handling and type validation

## Testing:
Created comprehensive test suites with 38 total test cases
Tests cover edge cases, different image types, null handling, consistency
All tests pass successfully with proper validation
Algorithm Characteristics:
Average Hash: Simple, fast, good for basic deduplication
Perceptual Hash: Robust to minor modifications (brightness, contrast)
Difference Hash: Efficient, good for detecting structural changes
Wavelet Hash: Robust to rotation and scaling operations
Crop-Resistant Hash: Robust to cropping and framing differences

Performance:
All algorithms implemented in Rust for optimal performance
Efficient memory usage with proper image resizing
Batch processing support for large datasets

## Related Issues

[Support various image hash functions for deduplication. #4889](https://github.com/Eventual-Inc/Daft/issues/4889)

## Checklist

- [] Documented in API Docs (if applicable)
- [] Documented in User Guide (if applicable)
- [] If adding a new documentation page, doc is added to `docs/mkdocs.yml` navigation
- [] Documentation builds and is formatted properly (tag @/ccmao1130 for docs review)

### CODE REVIEW

File: daft/expressions/expressions.py [hunk 1]
```diff
@@ -5279,6 +5279,123 @@ def mode(self) -> Expression:
         """
         return self.attribute("mode")
 
+    def average_hash(self) -> Expression:
```
Comment (universalmind303) [inline daft/expressions/expressions.py]: Instead of individual functions, could we just have a single parametrized `image_hash` that takes in these different algorithms. 


```py
col("img").image_hash(algorithm='average')

col("img").image_hash(algorithm='perceptual')
```

File: src/daft-image/src/functions/hash.rs
Comment (universalmind303) [inline src/daft-image/src/functions/hash.rs:1]: I'd suggest taking a look at encode/decode `src/daft-functions-binary/src/encode.rs` 
and serialize/deserialize `src/daft-functions-serde/src/deserialize.rs` for a better way to parametrize these. 

Instead I'd like to see a single `ImageHash` struct, that is parametrized over the hash algorithm


```rs
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub struct ImageHash;


#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum ImageHashAlgorithm {
    Average,
    CropResistant,
    Perceptual,
    // ...
}

#[derive(FunctionArgs)]
pub struct ImageHashArgs<T> {
    input: T,
    algorithm: ImageHashAlgorithm,
}



#[typetag::serde]
impl ScalarUDF for ImageHash {
// ...
}
```

Conceptually this is very similar pattern to what's established in those functions, Let's follow the same pattern here.

**Review by universalmind303 — CHANGES_REQUESTED** [review]

File: src/daft-image/src/functions/hash.rs
Comment (Kshitijmishradev) [reply src/daft-image/src/functions/hash.rs:1]: ok sure i would do that. Thank you!

File: daft/expressions/expressions.py (hunk 1 above)
Comment (Kshitijmishradev) [reply daft/expressions/expressions.py]: Hey, I just made the changes, made a single function, and removed all the individual functions. Will look into the rest of the failed test and work on it. Meanwhile, will you please check the changes?

File: daft/expressions/expressions.py [hunk 2]
```diff
@@ -5279,122 +5279,38 @@ def mode(self) -> Expression:
         """
         return self.attribute("mode")
 
-    def average_hash(self) -> Expression:
-        """Computes the average hash of an image for deduplication and similarity detection.
+    def hash(self, algorithm: str = "average") -> Expression:
```
Comment (universalmind303) [inline daft/expressions/expressions.py]: use a literal type instead of a str. it should provide better typing. 
```py
    def hash(self, algorithm: Literal["average", "perceptual", ...] = "average") -> Expression:
```

Comment (universalmind303) [reply daft/expressions/expressions.py]: one other small note. which is no fault of yours, but rather bad timing.  

It appears that you started this PR before this one https://github.com/Eventual-Inc/Daft/pull/5086. Which is deprecating the namespaces in favor of moving things into the `daft.functions` module.

so instead of `col("image").image.hash()` it becomes `image_hash(col("image"))`.

It should a rather minor change to move this in to the `daft.functions` module.

File: src/daft-image/src/ops.rs [hunk 3]
```diff
@@ -359,3 +546,355 @@ pub fn fixed_image_html_value(arr: &FixedShapeImageArray, idx: usize, truncate:
...
+    // Convert u8 to f64 and normalize
+    for y in 0..32 {
+        for x in 0..32 {
+            dct_matrix[y][x] = pixel_data[y * 32 + x] as f64;
+        }
+    }
+
+    // Apply 2D DCT
+    let mut dct_result = vec![vec![0.0f64; 32]; 32];
+    #[allow(clippy::needless_range_loop)]
```
Comment (universalmind303) [inline src/daft-image/src/ops.rs:621]: fix clippy instead of suppressing it.

**Comment by Kshitijmishradev** [comment]
@universalmind303 , sorry man, i made a mistake with the rebase, and the commit history is too messed up, so I made a new branch. Should we close this PR and create a new one, or would you like me to make the changes to this branch?  What do you say?

**Comment by universalmind303** [comment]
> sorry man, i made a mistake with the rebase, and the commit history is too messed up, so I made a new branch. Should we close this PR and create a new one, or would you like me to make the changes to this branch? What do you say?

yes that works! Sorry for the added confusion, just bad timing on that big refactor PR!

**Comment by Kshitijmishradev** [comment]
> > sorry man, i made a mistake with the rebase, and the commit history is too messed up, so I made a new branch. Should we close this PR and create a new one, or would you like me to make the changes to this branch? What do you say?
> 
> yes that works! Sorry for the added confusion, just bad timing on that big refactor PR!

Hey, I just created a New PR feat: implement unified image hash function with 5 algorithms #5227
